File: apps/visualizer/backend/api/jobs.py
import sqlite3

from database import (
    add_job_log,
    count_job_logs,
    count_jobs,
    create_job,
    get_active_jobs,
    get_job,
    list_jobs,
    update_job_field,
    update_job_status,
)
from library_db import JOB_TYPES_REQUIRING_CATALOG, describe_library_db
from utils.responses import success_paginated
from flask import Blueprint, current_app, jsonify, request

bp = Blueprint('jobs', __name__)


_DB_BUSY_MESSAGE = (
    'Database is temporarily busy — another operation is holding a write lock. '
    'Try again in a moment; if it keeps happening, check for duplicate backend processes.'
)


def _db_busy_response():
    """Return a 503 JSON response for transient SQLite lock/busy errors."""
    return jsonify({'error': _DB_BUSY_MESSAGE, 'code': 'db_busy'}), 503

@bp.route('/', methods=['GET'])
def list_all_jobs():
    status = request.args.get('status')
    limit = request.args.get('limit', default=50, type=int)
    offset = request.args.get('offset', default=0, type=int)
    limit = max(1, min(limit, 500))
    offset = max(0, offset)
    jobs = list_jobs(current_app.db, status=status, limit=limit, offset=offset)
    total = count_jobs(current_app.db, status=status)
    return success_paginated(jobs, total=total, offset=offset, limit=limit)

@bp.route('/', methods=['POST'])
def create_new_job():
    data = request.json

    if not data or 'type' not in data:
        return jsonify({'error': 'type is required'}), 400

    job_type = data['type']
    metadata = data.get('metadata', {})

    if job_type in JOB_TYPES_REQUIRING_CATALOG:
        status = describe_library_db()
        if not status.exists:
            return jsonify({
                'error': (
                    f"Cannot enqueue {job_type!r}: Lightroom catalog database is unavailable. "
                    f"{status.reason or ''}"
                ).strip(),
                'code': 'catalog_unavailable',
                'library_db': status.to_dict(),
            }), 422

    try:
        job_id = create_job(current_app.db, job_type, metadata)
        job = get_job(current_app.db, job_id)
    except sqlite3.OperationalError as e:
        if 'locked' in str(e).lower() or 'busy' in str(e).lower():
            return _db_busy_response()
        raise

    return jsonify(job), 201

@bp.route('/<job_id>', methods=['GET'])
def get_job_details(job_id):
    # Resolve the tail size **before** loading the job so ``get_job`` only
    # pulls the rows the client actually needs. The frontend's modal view
    # asks for ``logs_limit=200`` to ``1000`` in steady state; anything
    # bigger is a power-user override capped at 10k for safety.
    logs_limit_raw = request.args.get('logs_limit', type=int)
    if logs_limit_raw is None:
        effective_limit: int | None = None  # default tail
        include_all = False
    elif logs_limit_raw == 0:
        effective_limit = None
        include_all = True
    else:
        effective_limit = max(1, min(logs_limit_raw, 10_000))
        include_all = False

    job = get_job(
        current_app.db,
        job_id,
        logs_limit=effective_limit,
        include_all_logs=include_all,
    )
    if not job:
        return jsonify({'error': 'Job not found'}), 404

    # ``logs_total`` must reflect the *full* log count, not the tail length,
    # so the frontend can render "showing N of M" correctly.
    job['logs_total'] = count_job_logs(current_app.db, job_id)
    return jsonify(job)

@bp.route('/<job_id>', methods=['DELETE'])
def cancel_job(job_id):
    from app import get_job_runner, socketio

    try:
        job = get_job(current_app.db, job_id)

        if not job:
            return jsonify({'error': 'Job not found'}), 404

        if job['status'] == 'running':
            update_job_status(current_app.db, job_id, 'cancelled')
            r = get_job_runner()
            if r:
                r.signal_cancel(job_id)
            add_job_log(current_app.db, job_id, 'info', 'Cancel requested via API')
            updated = get_job(current_app.db, job_id)
            if socketio:
                socketio.emit('job_updated', updated)
            return jsonify(updated)

        if job['status'] == 'pending':
            update_job_status(current_app.db, job_id, 'cancelled')
            add_job_log(current_app.db, job_id, 'info', 'Cancel requested via API')
            updated = get_job(current_app.db, job_id)
            if socketio:
                socketio.emit('job_updated', updated)
            return jsonify(updated)

        # Idempotent cancel: a cancel against an already-terminal job returns
        # the current row with a hint instead of an error, so rapid clicks /
        # retries from the UI don't produce misleading 400 toasts. Terminal
        # states are ``cancelled``, ``completed``, and ``failed``.
        if job['status'] in ('cancelled', 'completed', 'failed'):
            r = get_job_runner()
            if job['status'] == 'cancelled' and r:
                # Make sure the worker's cancel flag is set even if we missed
                # it the first time (e.g. DB was flipped directly by another
                # process). Cheap no-op if the job isn't active.
                r.signal_cancel(job_id)
            return jsonify({
                **job,
                'cancel_noop': True,
                'cancel_noop_reason': f"Job is already {job['status']}",
            })

        return jsonify({'error': f"Cannot cancel job in status {job['status']!r}"}), 400
    except sqlite3.OperationalError as e:
        if 'locked' in str(e).lower() or 'busy' in str(e).lower():
            return _db_busy_response()
        raise

@bp.route('/<job_id>/retry', methods=['POST'])
def retry_job(job_id):
    try:
        job = get_job(current_app.db, job_id)

        if not job:
            return jsonify({'error': 'Job not found'}), 404

        if job['status'] not in ('failed', 'cancelled'):
            return jsonify({'error': 'Can only retry failed or cancelled jobs'}), 400

        update_job_status(current_app.db, job_id, 'pending', progress=0, current_step=None)
        update_job_field(current_app.db, job_id, 'error', None)
        current_app.db.execute(
            "UPDATE jobs SET error_severity = NULL WHERE id = ?",
            (job_id,),
        )
        current_app.db.commit()
        update_job_field(current_app.db, job_id, 'result', None)
        add_job_log(current_app.db, job_id, 'info', 'Job queued for retry')

        return jsonify(get_job(current_app.db, job_id))
    except sqlite3.OperationalError as e:
        if 'locked' in str(e).lower() or 'busy' in str(e).lower():
            return _db_busy_response()
        raise


@bp.route('/active', methods=['GET'])
def list_active_jobs():
    jobs = get_active_jobs(current_app.db)
    return jsonify(jobs)


@bp.route('/health', methods=['GET'])
def get_jobs_health():
    """Expose subsystem health so the UI can warn before users enqueue broken jobs.

    Returns library DB resolution info plus the job types that require it.
    Always 200 — ``library_db.exists == false`` is the signal to render a banner.
    """
    status = describe_library_db()
    return jsonify({
        'library_db': status.to_dict(),
        'jobs_requiring_catalog': sorted(JOB_TYPES_REQUIRING_CATALOG),
        'catalog_available': status.exists,
    })


# Staleness threshold for the processor heartbeat. The loop ticks every ~1s,
# so anything older than 15s is very likely a hang. Kept loose enough that a
# slow DB query or a long socket emit doesn't trigger false positives.
_PROCESSOR_STALE_AFTER_SECONDS = 15.0


@bp.route('/_processor_health', methods=['GET'])
def get_processor_health():
    """Diagnose the background job processor — was it started, is it ticking?

    This endpoint was added after an incident where a ``pending`` job sat
    untouched for minutes and there was no way to tell from the outside
    whether the processor thread was alive, stuck, or never started.

    Response shape
    --------------
    - ``running`` (bool): the module-level flag ``job_processor_running``.
    - ``started_at`` / ``last_iteration_at`` (float | null): unix timestamps.
    - ``last_iteration_age_seconds`` (float | null): derived convenience.
    - ``iterations_total`` (int): cheap proof-of-life counter.
    - ``current_job_id`` / ``current_job_started_at``: what the processor
      is actively working on, if anything.
    - ``pending_count`` (int): outstanding jobs in the DB; useful alongside
      ``current_job_id`` to spot "pending but not being picked up".
    - ``stale`` (bool): convenience flag — True if ``last_iteration_age_seconds``
      exceeds :data:`_PROCESSOR_STALE_AFTER_SECONDS`. Clients can treat it
      as the primary signal without reimplementing the threshold.
    - ``last_error`` (str | null): most recent exception string from the
      outer ``try`` in the processor loop.
    """
    import sys
    import time as _time

    # When ``app.py`` is started directly (``python app.py``), the running
    # process has the server registered under ``__main__`` and the blueprint
    # imports a *second* copy as ``app``. The processor thread lives in the
    # ``__main__`` module, so snapshotting via ``from app import ...`` returns
    # the dormant copy and this endpoint misreports the processor as dead.
    # We prefer ``__main__`` when it exposes the heartbeat, and fall back to
    # the ``app`` module (the path used by tests and by ``flask run``).
    candidate_modules = []
    main_mod = sys.modules.get('__main__')
    if main_mod is not None and hasattr(main_mod, 'get_processor_health_snapshot'):
        candidate_modules.append(main_mod)
    app_mod = sys.modules.get('app')
    if app_mod is not None and app_mod is not main_mod:
        candidate_modules.append(app_mod)

    # Choose the module whose processor has actually started. If none have,
    # fall back to the first candidate so we still return a well-formed
    # payload with ``running: false``.
    source_mod = None
    for mod in candidate_modules:
        if getattr(mod, 'job_processor_running', False):
            source_mod = mod
            break
    if source_mod is None and candidate_modules:
        source_mod = candidate_modules[0]

    if source_mod is None:
        # Pathological: neither module is importable. Import explicitly so
        # the error surfaces clearly rather than returning a blank payload.
        from app import get_processor_health_snapshot, job_processor_running
        snapshot = get_processor_health_snapshot()
        job_running = job_processor_running
    else:
        snapshot = source_mod.get_processor_health_snapshot()
        job_running = bool(getattr(source_mod, 'job_processor_running', False))
    last_iter = snapshot.get('last_iteration_at')
    age = None
    if isinstance(last_iter, (int, float)):
        age = max(0.0, _time.time() - float(last_iter))

    pending_count = count_jobs(current_app.db, status='pending')
    running_count = count_jobs(current_app.db, status='running')

    return jsonify({
        'running': job_running,
        'started_at': snapshot.get('started_at'),
        'last_iteration_at': last_iter,
        'last_iteration_age_seconds': age,
        'iterations_total': snapshot.get('iterations_total') or 0,
        'current_job_id': snapshot.get('current_job_id'),
        'current_job_started_at': snapshot.get('current_job_started_at'),
        'pending_count': pending_count,
        'running_count': running_count,
        'stale': (age is not None and age > _PROCESSOR_STALE_AFTER_SECONDS),
        'stale_threshold_seconds': _PROCESSOR_STALE_AFTER_SECONDS,
        'last_error': snapshot.get('last_error'),
    })

File: apps/visualizer/backend/src/api/jobs.ts
/**
 * Jobs API — list, create, inspect, cancel and retry background jobs.
 */
import { createRoute, z } from '@hono/zod-openapi';
import { config } from '../config.js';
import type { Db } from '../db/connection.js';
import {
  countJobLogs,
  countJobs,
  createJob,
  getActiveJobs,
  getJob,
  getJobLogStatsBulk,
  initJobsDb,
  listJobs,
  type JobRow,
} from '../db/jobs/jobs.js';
import { describeLibraryDb } from '../jobs/library-db.js';
import { catalogRequiringJobTypes } from '../jobs/registry.js';
import { getProcessorHealthSnapshot, signalCancel } from '../jobs/processor.js';
import { transitionCancel, transitionRetry } from '../jobs/transitions.js';
import { emitJobEvent } from '../websocket/events.js';
import { paginatedBody } from '../utils/responses.js';
import { createOpenApiApp } from './openapi.js';
import { jsonBody, withValidationError } from './route-helpers.js';
import { ErrorBody } from './schemas/errors.js';
import {
  CatalogUnavailableError,
  compactJobPayload,
  DbBusyError,
  Job,
  JobCreateRequest,
  JobListResponse,
  JobsHealth,
  JobsListResponse,
  JobsProcessorHealth,
} from './schemas/jobs.js';

export const jobsRoutes = createOpenApiApp();

const DB_BUSY_MESSAGE =
  'Database is temporarily busy — another operation is holding a write lock. ' +
  'Try again in a moment; if it keeps happening, check for duplicate backend processes.';

/**
 * Staleness threshold for the processor heartbeat.
 *
 * The loop ticks about once a second, so anything older than 15s is very likely a
 * hang — loose enough that a slow query or a long socket emit is not a false alarm.
 */
const PROCESSOR_STALE_AFTER_SECONDS = 15;

/** SQLite lock contention, which is a transient 503 rather than a 500. */
function isDbBusy(e: unknown): boolean {
  const code = (e as { code?: string } | null)?.code;
  if (code === 'SQLITE_BUSY' || code === 'SQLITE_LOCKED') return true;
  const message = String(e instanceof Error ? e.message : e).toLowerCase();
  return message.includes('locked') || message.includes('busy');
}

/**
 * Open `visualizer.db` for one operation, creating the schema if needed.
 *
 * `initJobsDb` rather than a read-only open even for GETs: this is the app's own
 * database, and a first run must not 404 the jobs page just because nothing has
 * enqueued a job yet.
 */
function withJobsDb<T>(fn: (db: Db) => T): T {
  const db = initJobsDb(config.VISUALIZER_DB);
  try {
    return fn(db);
  } finally {
    db.close();
  }
}

/**
 * Prepare a job row for a JSON response or a socket emit.
 *
 * Attaches the log summary counts from `job_logs` — the same numbers the list
 * endpoint reports, so a job looks identical however the client obtained it — then
 * compacts the checkpoint lists.
 */
function buildJobPayload(db: Db, job: JobRow, extra: Record<string, unknown> = {}): Job {
  const stats = getJobLogStatsBulk(db, [job.id]).get(job.id);
  const enriched = {
    ...job,
    logs_total: stats?.logs_total ?? job.logs_total ?? 0,
    warning_count: stats?.warning_count ?? job.warning_count ?? 0,
    error_count: stats?.error_count ?? job.error_count ?? 0,
    last_log_at: stats?.last_log_at ?? job.last_log_at ?? null,
    ...extra,
  };
  return compactJobPayload(enriched as unknown as Record<string, unknown>) as unknown as Job;
}

// --- list and create --------------------------------------------------------

const listRoute = createRoute({
  method: 'get',
  path: '/jobs/',
  tags: ['jobs'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(JobsListResponse) },
  }),
});

jobsRoutes.openapi(listRoute, (c) => {
  const status = c.req.query('status');
  // Unparseable query values fall back to the default.
  const toInt = (raw: string | undefined, fallback: number): number =>
    raw !== undefined && /^\s*[+-]?\d+\s*$/.test(raw) ? Number.parseInt(raw.trim(), 10) : fallback;
  const limit = Math.max(1, Math.min(toInt(c.req.query('limit'), 50), 500));
  const offset = Math.max(0, toInt(c.req.query('offset'), 0));

  return withJobsDb((db) => {
    const jobs = listJobs(db, { status: status ?? null, limit, offset }).map(
      (job) => compactJobPayload(job as unknown as Record<string, unknown>) as unknown as Job,
    );
    const total = countJobs(db, status ?? null);
    return c.json(paginatedBody(jobs, total, offset, limit), 200);
  });
});

const createRouteDef = createRoute({
  method: 'post',
  path: '/jobs/',
  tags: ['jobs'],
  request: { body: { content: jsonBody(JobCreateRequest) } },
  responses: withValidationError({
    201: { description: 'Created', content: jsonBody(Job) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    422: { description: 'Catalog unavailable', content: jsonBody(CatalogUnavailableError) },
    503: { description: 'Database busy', content: jsonBody(DbBusyError) },
  }),
});

jobsRoutes.openapi(createRouteDef, (c) => {
  // The handler's own 'type is required' 400 is unreachable — a missing `type` is 422
  // before this runs. The 400 stays declared because the contract lists it.
  const { type: jobType, metadata } = c.req.valid('json');

  // Refused at enqueue time rather than accepted and failed later: a queued job
  // that can never run is worse than a clear rejection, because the user has to
  // go find it in the list to learn why nothing happened.
  if (catalogRequiringJobTypes().has(jobType)) {
    const status = describeLibraryDb();
    if (!status.exists) {
      return c.json(
        {
          error: (
            `Cannot enqueue '${jobType}': Lightroom catalog database is unavailable. ` +
            `${status.reason ?? ''}`
          ).trim(),
          code: 'catalog_unavailable' as const,
          library_db: status,
        },
        422,
      );
    }
  }

  try {
    const payload = withJobsDb((db) => {
      const jobId = createJob(db, jobType, metadata);
      const job = getJob(db, jobId);
      if (!job) throw new Error('job vanished after insert');
      return buildJobPayload(db, job);
    });
    emitJobEvent('job_created', payload);
    return c.json(payload, 201);
  } catch (e) {
    if (isDbBusy(e)) return c.json({ error: DB_BUSY_MESSAGE, code: 'db_busy' as const }, 503);
    throw e;
  }
});

// --- active and health ------------------------------------------------------

const activeRoute = createRoute({
  method: 'get',
  path: '/jobs/active',
  tags: ['jobs'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(JobListResponse) },
  }),
});

jobsRoutes.openapi(activeRoute, (c) =>
  withJobsDb((db) =>
    c.json(
      getActiveJobs(db).map(
        (job) => compactJobPayload(job as unknown as Record<string, unknown>) as unknown as Job,
      ),
      200,
    ),
  ),
);

const healthRoute = createRoute({
  method: 'get',
  path: '/jobs/health',
  tags: ['jobs'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(JobsHealth) },
  }),
});

/**
 * Subsystem health, so the UI can warn before the user enqueues a broken job.
 *
 * Always 200: `library_db.exists === false` is itself the signal to render a
 * banner, and a non-200 would make the banner impossible to distinguish from the
 * endpoint being down.
 */
jobsRoutes.openapi(healthRoute, (c) => {
  const status = describeLibraryDb();
  return c.json(
    {
      library_db: status,
      jobs_requiring_catalog: [...catalogRequiringJobTypes()].sort(),
      catalog_available: status.exists,
    },
    200,
  );
});

const processorHealthRoute = createRoute({
  method: 'get',
  path: '/jobs/_processor_health',
  tags: ['jobs'],
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(JobsProcessorHealth) },
  }),
});

/**
 * Diagnose the background processor — was it started, is it ticking?
 *
 * Added after an incident where a `pending` job sat untouched with no way to tell
 * whether the processor was alive. `stale` encodes the threshold clients should read.
 */
jobsRoutes.openapi(processorHealthRoute, (c) => {
  const snapshot = getProcessorHealthSnapshot();
  const lastIter = snapshot.last_iteration_at;
  const age = lastIter === null ? null : Math.max(0, Date.now() / 1000 - lastIter);

  return withJobsDb((db) =>
    c.json(
      {
        running: snapshot.running,
        started_at: snapshot.started_at,
        last_iteration_at: lastIter,
        last_iteration_age_seconds: age,
        iterations_total: snapshot.iterations_total,
        current_job_id: snapshot.current_job_id,
        current_job_started_at: snapshot.current_job_started_at,
        pending_count: countJobs(db, 'pending'),
        running_count: countJobs(db, 'running'),
        stale: age !== null && age > PROCESSOR_STALE_AFTER_SECONDS,
        stale_threshold_seconds: PROCESSOR_STALE_AFTER_SECONDS,
        last_error: snapshot.last_error,
      },
      200,
    ),
  );
});

// --- one job ----------------------------------------------------------------
//
// Registered AFTER the static paths above. Hono matches in registration order, so
// `{job_id}` would otherwise swallow `active`, `health` and `_processor_health`.

const detailRoute = createRoute({
  method: 'get',
  path: '/jobs/{job_id}',
  tags: ['jobs'],
  request: { params: z.object({ job_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(Job) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
  }),
});

jobsRoutes.openapi(detailRoute, (c) => {
  const jobId = c.req.valid('param').job_id;

  // Resolved before the job is loaded, so `getJob` only pulls the rows the client
  // actually needs. The modal asks for 200–1000 in steady state; anything larger
  // is a power-user override, capped at 10k.
  const raw = c.req.query('logs_limit');
  const parsed = raw !== undefined && /^\s*[+-]?\d+\s*$/.test(raw)
    ? Number.parseInt(raw.trim(), 10)
    : null;
  let logsLimit: number | null = null;
  let includeAllLogs = false;
  if (parsed === 0) {
    includeAllLogs = true;
  } else if (parsed !== null) {
    logsLimit = Math.max(1, Math.min(parsed, 10_000));
  }

  return withJobsDb((db) => {
    const job = getJob(db, jobId, { logsLimit, includeAllLogs });
    if (!job) return c.json({ error: 'Job not found' }, 404);
    // The full count, not the tail length, so the UI can say "showing N of M".
    job.logs_total = countJobLogs(db, jobId);
    return c.json(buildJobPayload(db, job, { logs_total: job.logs_total }), 200);
  });
});

const cancelRoute = createRoute({
  method: 'delete',
  path: '/jobs/{job_id}',
  tags: ['jobs'],
  request: { params: z.object({ job_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(Job) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
    503: { description: 'Database busy', content: jsonBody(DbBusyError) },
  }),
});

jobsRoutes.openapi(cancelRoute, (c) => {
  const jobId = c.req.valid('param').job_id;
  try {
    const result = withJobsDb((db) => {
      const outcome = transitionCancel(db, jobId);
      if (outcome.edge === 'invalid' && outcome.job === null) {
        return { status: 404 as const, body: { error: 'Job not found' } };
      }
      if (outcome.edge === 'invalid') {
        return { status: 400 as const, body: { error: outcome.reason ?? 'Cannot cancel' } };
      }
      if (outcome.shouldSignalCancel) signalCancel(jobId);
      if (outcome.edge === 'noop') {
        return {
          status: 200 as const,
          emit: false,
          body: buildJobPayload(db, outcome.job!, {
            cancel_noop: true,
            cancel_noop_reason: outcome.reason,
          }),
        };
      }
      return { status: 200 as const, emit: true, body: buildJobPayload(db, outcome.job!) };
    });

    if (result.status === 404) return c.json(result.body as { error: string }, 404);
    if (result.status === 400) return c.json(result.body as { error: string }, 400);
    // A noop is not broadcast: nothing changed, and every connected client would
    // otherwise re-render the job for no reason.
    if (result.emit) emitJobEvent('job_updated', result.body as Job);
    return c.json(result.body as Job, 200);
  } catch (e) {
    if (isDbBusy(e)) return c.json({ error: DB_BUSY_MESSAGE, code: 'db_busy' as const }, 503);
    throw e;
  }
});

const retryRoute = createRoute({
  method: 'post',
  path: '/jobs/{job_id}/retry',
  tags: ['jobs'],
  request: { params: z.object({ job_id: z.string() }) },
  responses: withValidationError({
    200: { description: 'OK', content: jsonBody(Job) },
    400: { description: 'Bad Request', content: jsonBody(ErrorBody) },
    404: { description: 'Not Found', content: jsonBody(ErrorBody) },
    503: { description: 'Database busy', content: jsonBody(DbBusyError) },
  }),
});

jobsRoutes.openapi(retryRoute, (c) => {
  const jobId = c.req.valid('param').job_id;
  try {
    const result = withJobsDb((db) => {
      const outcome = transitionRetry(db, jobId);
      if (outcome.edge === 'invalid' && outcome.job === null) {
        return { status: 404 as const, body: { error: 'Job not found' } };
      }
      if (outcome.edge === 'invalid') {
        return { status: 400 as const, body: { error: outcome.reason ?? 'Cannot retry' } };
      }
      return { status: 200 as const, body: buildJobPayload(db, outcome.job!) };
    });
    if (result.status === 404) return c.json(result.body as { error: string }, 404);
    if (result.status === 400) return c.json(result.body as { error: string }, 400);
    return c.json(result.body as Job, 200);
  } catch (e) {
    if (isDbBusy(e)) return c.json({ error: DB_BUSY_MESSAGE, code: 'db_busy' as const }, 503);
    throw e;
  }
});
